File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2025/Y25D18.kt
package org.gristle.pdxpuzzles.everybodycodes.y2025

import org.gristle.pdxpuzzles.everybodycodes.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.blankSplit
import org.gristle.pdxpuzzles.utilities.parsing.getLongs

object Y25D18 : Day {
    data class Plant(val thickness: Long, val branches: Branches) {
        fun getEnergy(plants: List<Plant>, plantEnergy: MutableList<Long?>): Long = when (branches) {
            Free -> 1L
            is ToPlants -> {
                var energy = 0L
                for (branch in branches.value) {
                    val sourceEnergy = plantEnergy[branch.toPlant]
                        ?: let {
                            val branchEnergy = plants[branch.toPlant].getEnergy(plants, plantEnergy)
                            plantEnergy[branch.toPlant] = branchEnergy
                            branchEnergy
                        }
                    energy += sourceEnergy * branch.thickness
                }
                if (energy >= thickness) {
                    energy
                } else {
                    0
                }
            }
        }



        companion object {
            fun from(notes: String): Plant {
                val thickness = notes.getLongs().drop(1).first()
                val branches = Branches.from(notes.dropWhile { it != '\n' })
                return Plant(thickness, branches)
            }
        }
    }

    sealed interface Branches {
        companion object {
            fun from(value: String): Branches = if (value.startsWith("\n- f")) {
                Free
            } else {
                val branches = value.getLongs().chunked(2)
                    .map { (plant, thickness) ->
                        Branch(thickness, plant.toInt() - 1)
                    }.toList()
                ToPlants(branches)
            }
        }
    }

    data object Free : Branches

    @JvmInline
    value class ToPlants(val value: List<Branch>) : Branches

    data class Branch(val thickness: Long, val toPlant: Int)

    private fun getPlants(notes: String): List<Plant> = notes.blankSplit().map(Plant::from)

    override fun part1(input: String): Long {
        val plants = getPlants(input)
        val plantEnergy: MutableList<Long?> = MutableList(plants.size) { null }
        return plants.last().getEnergy(plants, plantEnergy)
    }

    override fun part2(input: String): Long {
        val (plantsStr, testCases) = input.split("\n\n\n")
        val plants = getPlants(plantsStr)

        return testCases.lineSequence().sumOf { line ->
            val iter = line.getLongs().iterator()
            val plantEnergy: MutableList<Long?> = MutableList(plants.size) {
                if (iter.hasNext()) iter.next() else null
            }
            plants.last().getEnergy(plants, plantEnergy)
        }
    }

    override fun part3(input: String): Long {
        val (plantsStr, testCases) = input.split("\n\n\n")
        val plants = getPlants(plantsStr)

        val plantEnergy: MutableList<Long?> = MutableList(plants.size) { null }

        val phase2Start = plants.indexOfFirst { (_, branches) -> branches is ToPlants }

        for (i in 0 until phase2Start) plantEnergy[i] = 0L

        val positiveBranchPlants = plants.asSequence()
            .mapNotNull { (_, branches) ->
                when (branches) {
                    Free -> null
                    is ToPlants -> branches.value
                }
            }.takeWhile { branches -> branches.size > 2 }
            .flatMap { branches ->
                branches
                    .filter { branch -> branch.thickness > 0 }
                    .map { branch -> branch.toPlant }
            }.distinct()
            .sorted()

        for (plantIdx in positiveBranchPlants) {
            plantEnergy[plantIdx] = 1L
        }

        val optimum = plants.last().getEnergy(plants, plantEnergy)

        return testCases.lineSequence().sumOf { line ->
            val iter = line.getLongs().iterator()
            val plantEnergy: MutableList<Long?> = MutableList(plants.size) {
                if (iter.hasNext()) iter.next() else null
            }
            val energy = plants.last().getEnergy(plants, plantEnergy)
            if (energy == 0L) {
                0L
            } else {
                optimum - energy
            }
        }
    }
}

fun main() = Day.runDay(Y25D18::class)

//    Quest 1: 2067316 (3ms)
//    Quest 2: 15481956620 (7ms)
//    Quest 3: 485271 (6ms)
//    Total time: 17ms

File: rust/everybody-codes/src/bin/2025/y25d18.rs
use everybody_codes::utilities::inputs::get_event_inputs;
use itertools::Itertools;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input<'a> = &'a str;
type Plants = Vec<Plant>;

#[derive(Debug)]
struct Plant {
    thickness: i64,
    branches: Branches,
}

impl Plant {
    fn get_energy(&self, plants: &[Plant], plant_energy: &mut Vec<Option<i64>>) -> i64 {
        match &self.branches {
            Branches::Free => 1,
            Branches::ToPlants(branches) => {
                let mut energy = 0;
                for branch in branches {
                    let branch_energy = plant_energy[branch.to_plant];
                    let source_energy = if let Some(branch_energy) = branch_energy {
                        branch_energy
                    } else {
                        let branch_energy = plants[branch.to_plant].get_energy(plants, plant_energy);
                        plant_energy[branch.to_plant] = Some(branch_energy);
                        branch_energy
                    };
                    energy += source_energy * branch.thickness;
                }
                if energy >= self.thickness {
                    energy
                } else {
                    0
                }
            },
        }
    }
}

impl From<&str> for Plant {
    fn from(description: &str) -> Self {
        let (plant, branches) = description.split_once("\n").unwrap();
        let thickness: i64 = plant.get_numbers().skip(1).next().unwrap();
        let branches = Branches::from(branches);
        Self { thickness, branches }
    }
}

#[derive(Debug, PartialEq, Eq)]
enum Branches {
    Free,
    ToPlants(Vec<Branch>)
}

impl From<&str> for Branches {
    fn from(value: &str) -> Self {
        if value.starts_with("- f") {
            Self::Free
        } else {
            let branches = value.get_numbers().tuples()
                .map(|(plant, thickness)| Branch { thickness, to_plant: plant as usize - 1 })
                .collect();
            Self::ToPlants(branches)
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
struct Branch {
    thickness: i64,
    to_plant: usize,
}

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_event_inputs(25, 18);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn get_plants(notes: Input) -> Plants {
    notes.split("\n\n").map(Plant::from).collect()
}

fn part1(notes: Input) -> i64 {
    let plants = get_plants(notes);
    let mut plant_energy: Vec<Option<i64>> = vec![None; plants.len()];
    plants.last().map(|plant| plant.get_energy(&plants, &mut plant_energy)).unwrap()
}

fn part2(notes: Input) -> i64 {
    let (plants, test_cases) = notes.split_once("\n\n\n").unwrap();
    let plants = get_plants(plants);
    let last = plants.last().unwrap();
    test_cases.lines()
        .map(|line| {
            let mut plant_energy: Vec<Option<i64>> = line.get_numbers()
                .map(|n| Some(n))
                .collect();
            plant_energy.resize(plants.len(), None);
            last.get_energy(&plants, &mut plant_energy)
        })
        .sum()
}

fn part3(notes: Input) -> i64 {
    let (plants, test_cases) = notes.split_once("\n\n\n").unwrap();
    let plants = get_plants(plants);
    let last = plants.last().unwrap();

    let mut plant_energy: Vec<Option<i64>> = vec![None; plants.len()];

    let phase_2_start = plants.iter()
        .position(|plant| {
            matches!(plant.branches, Branches::ToPlants(_))
        })
        .unwrap();
    for i in 0..phase_2_start {
        plant_energy[i] = Some(0);
    }
    
    let positive_branch_plants = plants.iter()
        .filter_map(|plant| {
            match &plant.branches {
                Branches::Free => None,
                Branches::ToPlants(branches) => Some(branches),
            }
        })
        .take_while(|branches| branches.len() > 2)
        .flat_map(|branches| {
            branches
                .iter()
                .filter(|branch| branch.thickness > 0)
                .map(|branch| branch.to_plant)
        })
        .unique()
        .sorted_unstable();

    for plant_idx in positive_branch_plants {
        plant_energy[plant_idx] = Some(1);
    }

    let optimum = last.get_energy(&plants, &mut plant_energy);

    test_cases.lines()
        .map(|line| {
            let mut plant_energy: Vec<Option<i64>> = line.get_numbers()
                .map(|n| Some(n))
                .collect();
            plant_energy.resize(plants.len(), None);
            let energy = last.get_energy(&plants, &mut plant_energy);
            if energy == 0 {
                0
            } else {
                optimum - energy
            }
        })
        .sum()
}

#[test]
fn default() {
    let (input1, input2, input3) = get_event_inputs(25, 18);
    assert_eq!(2067316, part1(&input1));
    assert_eq!(15481956620, part2(&input2));
    assert_eq!(485271, part3(&input3));
}

// Input parsed (44μs)
// 1. 2067316 (16μs)
// 2. 15481956620 (86μs)
// 3. 485271 (148μs)
// Total: 298μs
